libnetwork/resolvconf/resolvconf.go

```diff
@@ -58,32 +66,478 @@ func Path() string {
+	}
+	for _, ns := range rc.md.ExtNameServers {
+		var nss string
+		if ns.HostLoopback {
+			nss = "host(" + ns.Addr.String() + ")"
+		} else {
+			nss = ns.Addr.String()
+		}
+		s.ExtNameServers = append(s.ExtNameServers, nss)
+	}
+	if rc.md.NSOverride {
+		s.Overrides = append(s.Overrides, "nameservers")
+	}
+	if rc.md.SearchOverride {
+		s.Overrides = append(s.Overrides, "search")
+	}
+	if rc.md.OptionsOverride {
+		s.Overrides = append(s.Overrides, "options")
+	}
+
+	templText := `{{if .Comments}}{{with .Md.HeaderComment}}{{.}}
+
+{{end}}{{end}}{{range .NameServers -}}
+nameserver {{.}}
+{{end}}{{with .Search -}}
+search {{.}}
+{{end}}{{with .Options -}}
+options {{.}}
+{{end}}{{with .Other -}}
+{{.}}
+{{end}}{{if .Comments}}
+# Based on host file: '{{.Md.SourcePath}}'{{with .Md.Transform}} ({{.}}){{end}}
+{{if .Md.UsedDefaultNS -}}
+# Used default nameservers.
+{{end -}}
+{{with .ExtNameServers -}}
+# ExtServers: {{.}}
+{{end -}}
+# Overrides: {{.Overrides}}
+{{with .Md.NDotsFrom -}}
+# Option ndots from: {{.}}
+{{end -}}
+{{end -}}
+`
+	var buf bytes.Buffer
+	templ, err := template.New("summary").Parse(templText)
+	if err != nil {
+		return nil, errdefs.System(err)
+	}
+	if err := templ.Execute(&buf, s); err != nil {
+		return nil, errdefs.System(err)
+	}
+	return buf.Bytes(), nil
+}
+
+// WriteFile generates content and writes it to path. If hashPath is non-zero, it
+// also writes a file containing a hash of the content, to enable UserModified()
+// to determine whether the file has been modified.
+func (rc *ResolvConf) WriteFile(path, hashPath string, perm os.FileMode) error {
+	content, err := rc.Generate(true)
+	if err != nil {
+		return err
+	}
+
+	// Write the resolv.conf file - it's bind-mounted into the container, so can't
+	// move a temp file into place, just have to truncate and write it.
+	if err := os.WriteFile(path, content, perm); err != nil {
+		return errdefs.System(err)
+	}
+
+	// Write the hash file.
+	if hashPath != "" {
+		hashFile, err := ioutils.NewAtomicFileWriter(hashPath, perm)
+		if err != nil {
+			return errdefs.System(err)
+		}
+		defer hashFile.Close()
+
+		_, err = hashFile.Write(hashData(content))
+		if err != nil {
+			return errdefs.System(err)
+		}
+	}
+
+	return nil
+}
+
+// UserModified can be used to determine whether the resolv.conf file has been
+// modified since it was generated. It returns true if the files at rcPath and
+// rcHashPath both exist, and the hash file does not contain the hash of the
+// current rc file - else false.
+func UserModified(rcPath, rcHashPath string) bool {
+	currRC, err := os.ReadFile(rcPath)
+	if err != nil {
+		return !os.IsNotExist(err)
+	}
+
+	currRCHash, err := os.ReadFile(rcHashPath)
+	if err != nil {
+		return !os.IsNotExist(err)
+	}
+
+	return !bytes.Equal(hashData(currRC), currRCHash)
+}
+
+func readResolvConf(path string, reader io.Reader) (*ResolvConf, error) {
+	var rc ResolvConf
+	rc.md.SourcePath = path
```

Comment:
Done ... but still opening the file here if reader is nil, no need to clutter the caller with that?

Reply:
Forcing the caller to open the file and deal with any errors first is the point

Reply:
Hmm, ok. I can change it, but don't quite see the issue... they still get to deal with the error, it just comes from Parse instead of Open. The calling function ends up being simpler because it's only dealing with what it wants to deal with, rather than the mechanics of opening a file that it doesn't need the contents of?

Reply:
Haven't done that yet ... still can though, I expect there will be more updates to make! Interested to hear the reasoning.

Reply:
My reasoning was along the lines of not letting the caller (or code reviewers) overlook that a file is being opened and therefore that errors related to opening a file may be returned.
```go
rc, err := resolvconf.Parse(path, nil)
if err != nil {
    return nil, errors.New("malformed resolv.conf file") // WRONG
}
```
```go
rc, err := resolvconf.Parse(path, nil)
if err != nil {
    // File not found...right? WRONG
    rc = &resolvconf.ResolvConf{}
}
```
Separating file-open from parsing makes the above bugged code obviously incorrect by inspection.
```go
f, err := os.Open(path)
if err != nil {
    return nil, errors.New("malformed resolv.conf file")
}
defer f.Close()
rc, err := resolvconf.Parse(f, nil)
if err != nil {
    return nil, errors.new("file not found")
}
```

The tradeoff is more verbosity at the call site, and making it possible to forget `defer f.Close()`.

```go
var rc *resolvconf.ResolvConf
f, err := os.Open(path)
if err != nil {
    if !errors.Is(err, fs.ErrNotExist) {
        return nil, err
    }
    rc = &resolvconf.ResolvConf{}
    goto parsed
}
defer f.Close()
rc, err := resolvconf.Parse(f, "")
parsed:
```

Doesn't seem worth it, especially now that `resolvconf.Parse` is internal.

I do have one request, though: have two `ResolvConf` constructors, one which parses an io.Reader (`resolvconf.Parse`) and another which opens a file by path.
```go
func Load(path string) (*ResolvConf, error) {
    f, err := os.Open(path)
    if err != nil {
        return nil, err
    }
    return Parse(f, "")
}
```

If we're talking ergonomics, we could go even further:
```go
func Load(path string) (ResolvConf, error) {
    f, err := os.Open(path)
    if err != nil {
        return ResolvConf{}, err
    }
    return Parse(f, "")
}
```
If the constructors always return a `ResolvConf` value, zero-valued on error, the calling code can be simplified.
```go
rc, err := resolvconf.Load(path)
if err != nil && !errors.Is(err, fs.ErrNotExist) {
    return nil, err
}
// Proceed with rc, which might be zero-valued if path does not exist.
```
I like to return by-value in such situations as it is impossible by construction to return a nil pointer when you are not returning pointers.

(I realize this is a refinement of your original implementation... heh whoops.)

Reply:
It might also make sense to swap the argument order for `Parse`, to be more like [`os.NewFile`](https://pkg.go.dev/os#NewFile).

Reply:
Thanks Cory - that makes sense ... I'll go back to the two constructors as you suggest - I hadn't exported the one that accepted a Reader because it wasn't needed (yet) by libnetwork. Putting the new code in an internal package makes things a lot easier.